### PR TITLE
Upgrade Orion to TensorFlow 2.0

### PR DESCRIPTION
**Pull request by lcwong0928**
**Changes**
- Update packges in `setup.py`

**Removed**

**New**

### CODE REVIEW

File: orion/primitives/jsons/orion.primitives.tadgan.TadGAN.json [hunk 1]
```diff
@@ -1,356 +1,377 @@
...
+                "default": "tensorflow.keras.optimizers.Adam"
+            },
+            "input_shape": {
+                "type": "tuple",
+                "default": null
+            },
+            "latent_shape": {
+                "type": "tuple",
+                "default": null
+            },
```
Comment (sarahmish) [inline orion/primitives/jsons/orion.primitives.tadgan.TadGAN.json:65]: I propose removing `latent_shape` as a parameter, and keep `latent_dim` then construct `latent_shape = (latent_dim, output_dim)` in the `TadGAN` class.

Basically, I want to reduce the possibility of having two contradicting information, which can happen if the user specifies `latent_dim` and `latent_shape`.

File: orion/primitives/tadgan.py [hunk 2]
```diff
@@ -3,41 +3,61 @@
...
-            inputs[1] x_    predicted input
-        """
-        alpha = K.random_uniform((64, 1, 1))
-        return (alpha * inputs[0]) + ((1 - alpha) * inputs[1])
 
+def build_layer(layer: dict, hyperparameters: dict):
+    layer_class = import_object(layer['class'])
+    layer_kwargs = layer['parameters'].copy()
+    # TODO: Upgrade to using tf.keras.layers.Wrapper in mlprimitives.
+    if issubclass(layer_class, tf.keras.layers.Wrapper):
```
Comment (sarahmish) [inline orion/primitives/tadgan.py:33]: can you please open an [issue in MLPrimitives](https://github.com/MLBazaar/MLPrimitives/issues/) to mark this as an issue to update next release?

File: orion/primitives/tadgan.py [hunk 3]
```diff
@@ -3,41 +3,61 @@
...
     Args:
+        layers_encoder (list):
+            List containing layers of encoder.
+        layers_generator (list):
+            List containing layers of generator.
+        layers_critic_x (list):
+            List containing layers of critic_x.
+        layers_critic_z (list):
+            List containing layers of critic_z.
         shape (tuple):
```
Comment (sarahmish) [inline orion/primitives/tadgan.py:53]: I think this is now `input_shape`

File: orion/primitives/tadgan.py [hunk 4]
```diff
@@ -46,34 +66,74 @@ class TadGAN(object):
...
+        **hyperparameters (dict):
             Optional. Dictionary containing any additional inputs.
     """
 
+    def __init__(self, layers_encoder: list, layers_generator: list, layers_critic_x: list,
+                 layers_critic_z: list, optimizer: str, input_shape: Optional[tuple] = None,
+                 latent_shape: Optional[tuple] = None, target_shape: Optional[tuple] = None,
+                 latent_dim: int = 20, learning_rate: float = 0.005, epochs: int = 2000,
+                 batch_size: int = 64, iterations_critic: int = 5, shuffle: bool = True,
+                 callbacks: tuple = (), validation_ratio: float = 0.2,
```
Comment (sarahmish) [inline orion/primitives/tadgan.py:100]: update `epochs` to 50 and `validation_ration` to 0.0 to be consistent with the json file, also check the docstrings above as well.

File: orion/primitives/tadgan.py [hunk 5]
```diff
@@ -82,244 +142,317 @@ def __getstate__(self):
...
-        # to infer the shape
+        # Infers the shape.
         self.shape = self.shape or shape
         self.target_shape = self.target_shape or target_shape
-
-        self._setdefault(kwargs, 'generator_reshape_dim', length // 2)
-        self._setdefault(kwargs, 'generator_reshape_shape', (length // 2, 1))
-        self._setdefault(kwargs, 'encoder_reshape_shape', self.latent_shape)
-
+        self.latent_shape = self.latent_shape or (self.latent_dim, output_dim)
```
Comment (sarahmish) [inline orion/primitives/tadgan.py:183]: if you apply my suggestion of keeping `latent_dim` and removing `latent_shape` from the model parameters, this line becomes `self.latent_shape = (self.latent_dim, output_dim)`

File: orion/primitives/tadgan.py [hunk 6]
```diff
@@ -82,244 +142,317 @@ def __getstate__(self):
...
+        """Format losses into dictionary mapping metric names to their current value."""
+        output = dict()
+        if self.detailed_losses:
+            for i in range(len(losses)):
+                for j in range(len(losses[i])):
+                    output[LOSS_NAMES[i][j]] = losses[i][j]
+        else:
+            for i in range(len(losses)):
+                output[LOSS_NAMES[i][0]] = losses[i][0]
+        return output
```
Comment (sarahmish) [inline orion/primitives/tadgan.py:231]: minor: maybe add a space between the two lines for easy readability.

typically when you go back in indentation, we add a space.

File: orion/primitives/tadgan.py [hunk 7]
```diff
@@ -82,244 +142,317 @@ def __getstate__(self):
...
+    @tf.function
+    def _wasserstein_loss(y_true, y_pred):
+        return tf.reduce_mean(y_true * y_pred)
+
+    @staticmethod
+    @tf.function
+    def _gradient_penalty_loss(real, fake, critic: Model):
+        # Random weighted average to create interpolated signals.
+        batch_size = tf.shape(real)[0]
+        alpha = tf.random.uniform([batch_size, 1, 1], dtype=tf.float64)
```
Comment (sarahmish) [inline orion/primitives/tadgan.py:246]: curious if there is a reason to calculate `batch_size` rather than directly using `self.batch_size`?

File: orion/primitives/tadgan.py [hunk 8]
```diff
@@ -82,244 +142,317 @@ def __getstate__(self):
...
+    def fit(self, X: ndarray, y: Optional[ndarray] = None, **kwargs) -> None:
+        """Fit the TadGAN model.
 
         Args:
             X (ndarray):
-                N-dimensional array containing the input training sequences for the model.
+                N-dimensional array containing the input encoder sequences.
             y (ndarray):
-                N-dimensional array containing the target sequences we want to reconstruct.
+                Optional. N-dimensional array containing the input critic x sequences.
```
Comment (sarahmish) [inline orion/primitives/tadgan.py:395]: can you verify these docstrings? for `y` I think it's difficult for the user to understand what is critic x. I would rather keep it as before `target sequences the model is reconstructing`.

File: orion/primitives/tadgan.py [hunk 9]
```diff
@@ -82,244 +142,317 @@ def __getstate__(self):
...
+            **kwargs (dict):
+                Optional. Additional inputs.
         """
-        if y is None:
-            y = X.copy()  # reconstruct the same input
-
-        if not self._fitted:
-            self._augment_hyperparameters(X, y, kwargs)
+        # Infer dimensions and compile model.
+        y = X.copy() if y is None else y
```
Comment (sarahmish) [inline orion/primitives/tadgan.py:400]: is `y = y or X.copy()` more concise? or I would split it into multiple lines

```python3
if y is None:
    y = X.copy()

**Review by sarahmish — COMMENTED** [review]
Excellent! 

I have one proposed change, remove either `latent_shape` or `latent_dim` and construct on from the other.

File: orion/pipelines/verified/tadgan/tadgan_gpu.json [hunk 10]
```diff
@@ -32,110 +32,125 @@
...
-            "input_shape":[100, 1],
-            "target_shape": [100, 1],
             "layers_encoder": [
                 {
-                    "class": "keras.layers.Bidirectional",
+                    "class": "tensorflow.keras.layers.Bidirectional",
                     "parameters": {
                         "layer": {
-                            "class": "keras.layers.CuDNNLSTM",
+                            "class": "tensorflow.keras.layers.LSTM",
```
[the remark this answers is not in the document]
Comment (lcwong0928) [reply orion/pipelines/verified/tadgan/tadgan_gpu.json:40]: The difference between the two files is the `recurrent_dropout`. The original model uses a recurrent dropout rate of 0.0 but this violates the requirements to train LSTM layers on GPU:

Source:
https://www.tensorflow.org/api_docs/python/tf/keras/layers/LSTM

[the remark this answers is not in the document]
Comment (sarahmish) [reply orion/pipelines/verified/tadgan/tadgan_gpu.json:40]: I see, that sounds good then, then I would rename the file to `tadgan_without_dropout_gpu.json` or something similar to make it clear that is the main difference!

[the remark this answers is not in the document]
Comment (lcwong0928) [reply orion/pipelines/verified/tadgan/tadgan_gpu.json:40]: Renamed

File: orion/primitives/jsons/orion.primitives.tadgan.TadGAN.json (hunk 1 above)
Comment (lcwong0928) [reply orion/primitives/jsons/orion.primitives.tadgan.TadGAN.json:65]: Removed latent shape

File: orion/primitives/tadgan.py (hunk 3 above)
Comment (lcwong0928) [reply orion/primitives/tadgan.py:53]: updated

File: orion/primitives/tadgan.py (hunk 4 above)
Comment (lcwong0928) [reply orion/primitives/tadgan.py:100]: updated

File: orion/primitives/tadgan.py (hunk 7 above)
Comment (lcwong0928) [reply orion/primitives/tadgan.py:246]: This is for the edge case if the batch is less than `batch_size`, e.g. the leftover samples

File: orion/primitives/tadgan.py (hunk 6 above)
Comment (lcwong0928) [reply orion/primitives/tadgan.py:231]: Added space!

File: orion/primitives/tadgan.py (hunk 8 above)
Comment (lcwong0928) [reply orion/primitives/tadgan.py:395]: Reverted

File: orion/primitives/tadgan.py [hunk 11]
```diff
@@ -82,244 +139,321 @@ def __getstate__(self):
...
 
-    def predict(self, X, y=None):
-        """Predict values using the initialized object.
+    def predict(self, X: ndarray, y: Optional[ndarray] = None) -> tuple:
+        """Predict using TadGAN model.
 
         Args:
             X (ndarray):
-                N-dimensional array containing the input sequences for the model.
+                N-dimensional array containing the input training sequences for the model.
```
Comment (sarahmish) [inline orion/primitives/tadgan.py:438]: These are _inference_ sequences rather than training sequences.

**Review by sarahmish — COMMENTED** [review]
LGTM @lcwong0928! 

Couple of docstrings comments, but other than that, this is ready to be merged once tests are passing.

File: orion/pipelines/verified/tadgan/tadgan_without_dropout_gpu.json [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+        },
+        "sklearn.preprocessing.MinMaxScaler#1": {
+            "feature_range": [
+                -1,
+                1
+            ]
+        },
+        "mlprimitives.custom.timeseries_preprocessing.rolling_window_sequences#1": {
+            "target_column": 0,
+            "window_size": 100,
```
Comment (dyuliu) [inline orion/pipelines/verified/tadgan/tadgan_without_dropout_gpu.json:26]: reading this I just realized there is no easy guarantee some hyperparameters across different primitives share the same values. 

I am wondering whether mlblock supports the following feature or not?
```json
{
"mlprimitives.custom.timeseries_preprocessing.rolling_window_sequences#1": {
            "target_column": 0,
            "window_size": "orion.primitives.tadgan.TadGAN#1.input_shape",
            "target_size": 1
},
"orion.primitives.tadgan.TadGAN#1": {
            "epochs": 35,
            "input_shape": [100, 1],
            "...": "..."
}

```

File: orion/pipelines/verified/tadgan/tadgan_without_dropout_gpu.json [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+    "output_names": {
+        "orion.primitives.tadgan.TadGAN#1": {
+            "y": "y_hat"
+        },
+        "mlprimitives.custom.timeseries_preprocessing.rolling_window_sequences#1": {
+            "index": "X_index",
+            "target_index": "y_index"
+        }
+    }
+}
```
Comment (dyuliu) [inline orion/pipelines/verified/tadgan/tadgan_without_dropout_gpu.json:173]: why not just name this file as `tadgan_gpu`?

File: orion/primitives/jsons/orion.primitives.tadgan.TadGAN.json [hunk 14]
```diff
@@ -1,356 +1,373 @@
...
-                      "class": "keras.layers.Dense",
-                      "parameters": {
-                          "units": 1
-                      }
-                  }
-              ]
-          }
-      }
-  }
+    "name": "orion.primitives.tadgan.TadGAN",
```
Comment (dyuliu) [inline orion/primitives/jsons/orion.primitives.tadgan.TadGAN.json:2]: Can you double check the style — 2 spaces or 4 spaces — for json file?

File: orion/primitives/jsons/orion.primitives.tadgan.TadGAN.json [hunk 15]
```diff
@@ -1,356 +1,373 @@
...
+    },
+    "hyperparameters": {
+        "fixed": {
+            "optimizer": {
+                "type": "str",
+                "default": "tensorflow.keras.optimizers.Adam"
+            },
+            "input_shape": {
+                "type": "tuple",
+                "default": null
```
Comment (dyuliu) [inline orion/primitives/jsons/orion.primitives.tadgan.TadGAN.json:60]: null? If this variable is not used, it should be deleted. The same for other null variables in fixed section.

**Comment by dyuliu** [comment]
Remember to fix style check issues.

File: orion/pipelines/verified/tadgan/tadgan_without_dropout_gpu.json (hunk 12 above)
Comment (lcwong0928) [reply orion/pipelines/verified/tadgan/tadgan_without_dropout_gpu.json:26]: This feature is not currently supported as far as I know.

File: orion/primitives/jsons/orion.primitives.tadgan.TadGAN.json (hunk 15 above)
Comment (lcwong0928) [reply orion/primitives/jsons/orion.primitives.tadgan.TadGAN.json:60]: It defaults to null so that `_augment_hyperparameters` could infer it if it is null.

File: orion/primitives/jsons/orion.primitives.tadgan.TadGAN.json (hunk 14 above)
Comment (lcwong0928) [reply orion/primitives/jsons/orion.primitives.tadgan.TadGAN.json:2]: `lstm_autoencoder.json` and `lstm_dynamic_threshold.json` both seem to use 4 spaces for tabs

**Comment by dyuliu** [comment]
Should figure out why 2.0 implementation eats more memory, leading to the failure of the tutorial tests

File: orion/pipelines/verified/tadgan/tadgan_without_dropout_gpu.json (hunk 12 above)
Comment (lcwong0928) [reply orion/pipelines/verified/tadgan/tadgan_without_dropout_gpu.json:26]: Documented issue in sintel-dev channel.

File: orion/primitives/tadgan.py [hunk 16]
```diff
@@ -82,235 +127,254 @@ def __getstate__(self):
...
 
-                state[network] = keras.models.load_model(fd.name)
+                state[network] = tf.keras.models.load_model(fd.name)
 
         self.__dict__ = state
 
-    def _build_model(self, hyperparameters, layers, input_shape):
-        x = Input(shape=input_shape)
-        model = keras.models.Sequential()
+    @classmethod
```
Comment (sarahmish) [inline orion/primitives/tadgan.py:146]: Is there a reason this is a class method? I don't have a preference just wondering about the logic.

File: orion/primitives/tadgan.py [hunk 17]
```diff
@@ -82,235 +127,254 @@ def __getstate__(self):
...
+
+            # Calculate the gradients w.r.t to this interpolated signal.
+            grads = gp_tape.gradient(validity_interpolated, [interpolated])[0]
+
+            # Calculate the norm of the gradients.
+            norm = tf.sqrt(tf.reduce_sum(tf.square(grads), axis=np.arange(1, len(grads.shape))))
+            gradient_penalty = tf.reduce_mean((norm - 1.0) ** 2)
+            return gradient_penalty
+
+        return _gradient_penalty_loss
```
Comment (sarahmish) [inline orion/primitives/tadgan.py:225]: minor: for static methods, can we place them directly before `__init__` function? I think that's the proper convention of where to place it in a class

File: orion/primitives/tadgan.py [hunk 18]
```diff
@@ -82,235 +127,254 @@ def __getstate__(self):
...
+            hyperparameters, self.layers_critic_z, self.critic_z_input_shape, name='critic_z')
+
+        # Optimizers
+        self.critic_x_optimizer = import_object(self.optimizer)(self.learning_rate)
+        self.critic_z_optimizer = import_object(self.optimizer)(self.learning_rate)
+        self.encoder_generator_optimizer = import_object(self.optimizer)(self.learning_rate)
+
+        x = tf.keras.Input(shape=self.shape)
+        y = tf.keras.Input(shape=self.target_shape)
+        z = tf.keras.Input(shape=(self.latent_dim, 1))
```
Comment (sarahmish) [inline orion/primitives/tadgan.py:248]: why not use `self.latent_shape` directly? 

There is a mismatch here between `(self.latent_dim, 1)` and `self.latent_shape` which is `(self.latent_dim, output_dim)`. I'm assuming that the latent shape is the correct dimension based on the architecture of critic z and the output of the encoder.

File: orion/primitives/tadgan.py [hunk 19]
```diff
@@ -82,235 +127,254 @@ def __getstate__(self):
...
 
-    def fit(self, X, y=None, **kwargs):
-        """Fit the TadGAN.
+    def fit(self, X: ndarray, y: Optional[ndarray] = None, **kwargs) -> None:
+        """Fit the TadGAN model.
 
         Args:
             X (ndarray):
-                N-dimensional array containing the input training sequences for the model.
+                N-dimensional array containing the input inference sequences for the model.
```
Comment (sarahmish) [inline orion/primitives/tadgan.py:342]: `X` here is training sequences

**Review by sarahmish — COMMENTED** [review]
The new implementation looks good @lcwong0928!
There are a few comments. Also, can you please open an issue reporting the runtime differences between this version and the TF 1.0 version?

File: orion/primitives/tadgan.py (hunk 19 above)
Comment (lcwong0928) [reply orion/primitives/tadgan.py:342]: Updated the docstrings

File: orion/primitives/tadgan.py (hunk 17 above)
Comment (lcwong0928) [reply orion/primitives/tadgan.py:225]: Moved stated methods to before `__init__`

**Comment by sarahmish** [comment]
Scores are posted in https://github.com/sintel-dev/Orion/issues/161#issuecomment-1253974462

**Review by dyuliu — APPROVED** [review]
